aips/aip-x.md

```diff
@@ -0,0 +1,707 @@
+import {
+  deserializeSignInPublicKey,
+  deserializeSignInSignature,
+  verifySignInMessage,
+} from "@aptos-labs/siwa";
+
+const app = new Hono();
+
+app.post("/auth/siwa/verify", (c) => {
+  const { type, signature, message, publicKey } = c.req.json();
+
+  const input = JSON.parse(getCookie(c, "siwa-input"));
+
+  const verification = verifySignInMessage(input, {
+    publicKey: deserializeSignInPublicKey(type, publicKey),
+    signature: deserializeSignInSignature(type, signature),
+    message,
+  });
+
+  if (!verification.valid) return c.json({ errors: verification.errors }, 400);
+
+  // ... create and store a session for the user
+
+  return c.json({ data: verification });
+});
+```
+
+4. Once the verification endpoint has been created and the wallet successfully responds to the application, verify the response with the new verification endpoint.
+
+```tsx
+export default function SignInPage() {
+  const { signIn } = useWallet();
+
+  const handleSignIn = async () => {
+    const input = await fetch("/auth/siwa");
+
+    const response: AptosSignInOutput = await signIn(input);
+
+    const verification = await fetch("/auth/siwa/verify", {
+      method: "POST",
+      body: JSON.stringify({
+        type: response.message.type,
+        signature: response.signature.bcsToHex().toString(),
+        message: response.message,
+        publicKey: response.publicKey.bcsToHex().toString(),
+      }),
+    });
+
+    if (!verification.ok) throw new Error("Failed to verify message");
+
+    // ... the user is signed in, redirect to application
+  };
+
+  // ...
+}
+```
+
+## Testing
+
+The new `@aptos-labs/siwa` package will have its utilities tested to ensure that validation is done correctly. Coverage is most important here. In addition, signature verification methods will be used from the `@aptos-labs/ts-sdk`, it's important that public key verification methods are supported and tested.
+
+## Risks and Drawbacks
+
+The ecosystem is responsible for adopting the new standard as an alternative to the `connect` + `signMessage` flow. If applications/wallets do not properly implement the `signIn` flow, they may suffer vulnerabilities that can lead to potential phishing, replay, or forgery attacks. If vulnerabilities are found within _Aptos Labs_ or ecosystem utilities, it’s important to patch and deprecate them as soon as possible. Applications that accept the `connect` + `signMessage` flow may still be vulnerable to security attacks.
+
+## Security Considerations
+
+It is important that signature verification methods are implemented correctly through the `@aptos-labs/ts-sdk`. If not correct, applications and wallets utilizing these methods will be vulnerable to attacks. Additional utilities provided from the `@aptos-labs/siwa` must be well tested to avoid attacks.
+
+The attack vectors that must be accounted for are the following:
+
+- Phishing attacks — A user must have visibility on signing a message that is for a different application.
```

Comment:
Such a phishing prompt should never be shown to the user. (So, really, the user should not have any visibility into this.)